### PR TITLE
implement Bloom filter IoC blacklisting

### LINKED ISSUE #41: Identify and warn on potentially problematic Bloom filters
It is sometimes desirable to reject a Bloom filter if it contains elements that, when used as an indicator, would lead to potentially excessive false positive alerts, e.g. `/`. It should be possible to specify such values in a blacklist.

### PR DESCRIPTION
**Pull request by satta**
Closes #41.

### CODE REVIEW

File: processing/bloom_handler.go [hunk 1]
```diff
@@ -131,17 +132,29 @@ func MakeBloomHandler(iocBloom *bloom.BloomFilter,
 		ForwardHandler:    forwardHandler,
 		DoForwardAlert:    (util.ForwardAllEvents || util.AllowType("alert")),
 		AlertPrefix:       alertPrefix,
+		BlacklistIOCs:     make([]string, 0),
```
Comment (hillu) [inline processing/bloom_handler.go]: the `, 0` is not needed, is it?

File: processing/bloom_handler.go [hunk 2]
```diff
@@ -156,7 +169,15 @@ func MakeBloomHandlerFromFile(bloomFilename string, compressed bool,
 			return nil, err
 		}
 	}
+	log.WithFields(log.Fields{
+		"domain": "bloom",
+	}).Infof("loading Bloom filter '%s'", bloomFilename)
```
Comment (hillu) [inline processing/bloom_handler.go]: May want to move this up towards beginning of the function, before any error path

File: processing/bloom_handler.go [hunk 3]
```diff
@@ -156,7 +169,15 @@ func MakeBloomHandlerFromFile(bloomFilename string, compressed bool,
...
 	bh := MakeBloomHandler(iocBloom, databaseChan, forwardHandler, alertPrefix)
+	err = checkBlacklisted(iocBloom, blacklistIOCs)
+	if err != nil {
+		return nil, err
+	}
```
Comment (hillu) [inline processing/bloom_handler.go]: You may want to use `if err := checkBlacklisted(...); err != nil { ... }` ... and move this block up before initializing `bh`.

File: processing/bloom_handler.go [hunk 4]
```diff
@@ -181,6 +202,10 @@ func (a *BloomHandler) Reload() error {
 			return err
 		}
 	}
+	err = checkBlacklisted(iocBloom, a.BlacklistIOCs)
+	if err != nil {
+		return err
+	}
```
Comment (hillu) [inline processing/bloom_handler.go]: `if err := ... ; err != nil { ... }` -- see above

File: processing/bloom_handler.go (hunk 1 above)
Comment (hillu) [inline processing/bloom_handler.go]: Leaving this at its default value, nil, should be ok. The only place where BlacklistIOCs is accessed seems to be a `for ... range` loop in `checkBlacklisted`.

File: processing/bloom_handler.go (hunk 3 above)
Comment (satta) [reply processing/bloom_handler.go]: True (moving up). But IMHO the way the code is written (assignment and check on two lines) is more idiomatic looking that the code of the rest of the project.

**Comment by hillu** [comment]
Please consider renaming the feature. I'd expect a blacklist to mean something different than what has been implemented here. The filter might be accepted and events matching the blacklist might not be processed. I'd certainly not expect the filter to be rejected.

"Forbidden IoCs" might be a better term.

**Comment by satta** [comment]
Agreed. It might even be possible to use a hybrid approach:
- check on BLF load for 'blacklisted' IoCs, log warning if any are present but do not reject the filter completely
- also do not generate alerts for any hits to 'blacklisted' IoCs -- this obviously has a processing overhead but can help making sure that up-to-date IoCs are always used, as long as they are not blacklisted.

**Comment by satta** [comment]
@hillu I pushed some changes implementing 'true' blacklisting.

**Comment by hillu** [comment]
Looks good.

**Comment by hillu** [comment]
Looks good.

**Comment by satta** [comment]
Then merge at will s'il vous plait :)

**Pull request merged**